code/modules/asset_cache/asset_pack.dm

```diff
@@ -0,0 +1,201 @@
+		return FALSE
+	loaded_urls = active_transport.load_asset_items_from_pack(src)
+
+/**
+ * checks if we're pushed to a transport
+ */
+/datum/asset_pack/proc/is_pushed_to_transport()
+	return !isnull(loaded_urls)
+
+/**
+ * generates whatever needs to be generated
+ *
+ * this must be idempotent and hopefully efficient,
+ * because this is called again if the transport is changed mid-round.
+ *
+ * @return return value is passed to register().
+ */
+/datum/asset_pack/proc/generate()
+	return
+
+/**
+ * packs a list of filename = file's into /datum/asset_item's
+ *
+ * @return filename = /datum/asset_item
+ */
+/datum/asset_pack/proc/pack(list/files)
+	. = list()
+	for(var/filename in files)
+		var/file = files[filename]
+		var/datum/asset_item/item = new(filename, file, do_not_mangle, absolute, FALSE)
+		.[filename] = item
+	if(do_not_separate && !absolute) // if absolute, it's pointless to generate namespace as we'll be in the same folder anyways.
+		// generate their namespace and assign
+		var/namespace = generate_namespace(.)
+		for(var/datum/asset_item/item as anything in .)
+			item.namespace_id = namespace
+
+/**
+ * things must be packed!
+ */
+/datum/asset_pack/proc/generate_namespace(list/datum/asset_item/items)
+	var/list/joining = list()
+	for(var/datum/asset_item/item as anything in items)
+		joining += item.hash
+	return md5(jointext(joining, "-"))
+
+/**
+ * registers our content to the asset system's transport
+ *
+ * returns a list of filenames associated to files,
+ * because that's what the asset items / cache is actually dealing with.
+ *
+ * all filenames must be *globally unique*.
+ * while this is technically able to be circumvented on the modern asset loading system
+ * used by tgui windows, it is nonetheless still kept and enforced
+ * to make debugging / development less awful.
+ *
+ * any temporary files should go into ""
+ *
+ * @params
+ * * generation - output of generate().
+ *
+ * @return list("<filename>" = file, ...)
+ */
+/datum/asset_pack/proc/register(generation)
+	return list()
+
+/datum/asset_pack/proc/get_url(filename)
+	. = loaded_urls[filename]
+	if(isnull(.))
+		CRASH("failed to get url for [filename] on [type] ([id || "compile-time"]). something is terribly wrong!")
+
+//* below is tg stuff *//
+
+// todo: i honestly don't yet know what this does / why we use custom loaders
+//       but when i do i'll probably touch up on it ~silicons
+
+/datum/asset_pack
+
+	var/cached_serialized_url_mappings
+	var/cached_serialized_url_mappings_transport_type
+
+/datum/asset_pack/proc/get_url_mappings()
+	return loaded_urls
+
+/// Returns a cached tgui message of URL mappings
+/datum/asset_pack/proc/get_serialized_url_mappings()
+	if (isnull(cached_serialized_url_mappings) || cached_serialized_url_mappings_transport_type != SSassets.transport.type)
+		cached_serialized_url_mappings = TGUI_CREATE_MESSAGE("asset/mappings", get_url_mappings())
+		cached_serialized_url_mappings_transport_type = SSassets.transport.type
+
```

Comment:
this basically tells tgui to call tgui.html's loadcss/loadjs methods

as far as i can tell it's just a Better Way To Dynamically Load Things 
at some point i'll probably be touching up on tgui and how its assets are done when i do the 516 tgui refresh